draft-banks-quic-performance.md

```diff
@@ -110,6 +113,64 @@ When a client uses a unidirectional stream to request a response payload from
 the server, the server opens a new unidirectional stream to send the requested
 data.  If no data is requested by the client, the server need take no action.
 
+## Unreliable Datagrams
+
+For endpoints that support the QUIC Unreliable Datagram extension, datagrams may
+also be used.  The datagrams are used similarly, but with a few minor differences
+to account for their unreliable nature:
```

Comment:
I made a bunch of comments and then realised that this entire section describes the request/response pattern in a pretty obtuse way. I'm still not sure if the client sends payload data in DATAGRAMS or not (I think not, and that seems like an artificial constraint). I don't think saying similar but different helps, so I suggest just stating it all explicitly. 

If I modelled this as `PERF_DATAGRAM` frames, I think the client and server send different frames? E.g. client sends {ud64: flow_id, u64: response_length} and server sends {u64: flow_id, array: response bytes}. Is that correct? If so, that seems to be very restrictive and prevents designing tests that mimic datagram flows.

A potential better option here is to say that the PERF_DATAGRAM frame (like the H3_DATAGRAM frame is an overlay on the DATAGRAM frame) has a  symmetrical format across client and server. The client only uses even flow IDs, the server uses only odd IDs. Then, the first time that an endpoint sees a valid flow ID, it MUST begin to respond by sending all of the bytes, carried in datagrams, up to the maximum length. It's not clear if that length is invariant - is a flow response_length allowed to change across different DATAGRAMs for the same flow? 

A single datagram frame type allows an endpoint to send as much unsolicited data as it chooses. An alternative model would be to have a "request frame" and "response frame", so that endpoints get some more control over the data the will receive - this seems a bit overkill to me.

Reply:
In general, the perf protocol is 100% client driven. I don't think that's explicitly stated, but it is a primary goal. I need to be able to have a general (ideally simple) server side logic that is still extensible enough to allow for a client to trigger most useful types of traffic patterns.

My goal for datagrams is to follow a similar pattern to streams, but it has two major differences.

1. We add an application layer flow ID to distinguish parallel "streams" of unreliable data. This flow ID and the response length (which is the same on all datagrams in that flow) must be repeated on each datagram sent in the flow to account for possible packet loss.
2. The application layer has to segment the total payload sent for each flow manually into separate datagrams.

As far as the framing goes, this then means the client sends {flow_id, response_length, payload} per datagram. The server then sends {flow_id, payload} per datagram. You could call this a frame type, but I think that might be a bit overkill here. Also, the reason the format is not symmetrical is because the server always responds to the client. It never sends anything unsolicited.

Are we on the same page or not here? I'm still not sure.

Reply:
@lpardue any additional comments here? Otherwise I will merge in another day or so, as is.